state/watcher/txnwatcher.go

```diff
@@ -388,5 +472,13 @@ func (w *TxnWatcher) sync() (bool, error) {
 	if err := iter.Close(); err != nil {
 		return false, errors.Annotate(err, "watcher iteration error")
 	}
+	// If we have exited the iterator without consuming all the txns since we
+	// last synced, or the collection has looped that's an issue.
+	if lastId != nil && lastSeenId != lastId {
```

Comment:
So this is potentially something that we want very good testing of.
AFAICT the old code was potentially very dangerous in that events could go missing. But this new code could potentially cause a lot *more* desyncs in the wild.
We should probably try to do some scale testing against it, as well as making sure it has good SolQA coverage and we are checking for these warnings/errors in the log.
This is the sort of error that I would probably like to make fatal in SolQA